solidity/contracts/lib/registry.sol

```diff
@@ -15,36 +15,66 @@
 // limitations under the License.
 pragma solidity ^0.8.20;
 
-import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {SmtLib} from "@iden3/contracts/lib/SmtLib.sol";
+import {PoseidonUnit2L, PoseidonUnit3L} from "@iden3/contracts/lib/Poseidon.sol";
+import {Commonlib} from "./common.sol";
+import "hardhat/console.sol";
 
-/// @title A sample on-chain implementation of an account mapping between Ethereum addresses and BabyJubjub public keys
+uint256 constant MAX_SMT_DEPTH = 64;
+
+/// @title A sample on-chain implementation of a KYC registry
+/// @dev The registry uses a Sparse Merkle Tree to store the
+///   Zeto accounts (Babyjubjub keys), so that transaction
+///   submitters can generate proofs of membership for the
+///   accounts in a privacy-preserving manner.
 /// @author Kaleido, Inc.
-contract Registry is Ownable {
-    mapping(address => uint256[2]) private publicKeys;
+abstract contract Registry {
+    SmtLib.Data internal _publicKeysTree;
+    using SmtLib for SmtLib.Data;
 
-    error AlreadyRegistered(address addr);
+    error AlreadyRegistered(uint256[2]);
 
-    constructor() Ownable(msg.sender) {}
+    constructor() {
+        _publicKeysTree.initialize(MAX_SMT_DEPTH);
+    }
 
-    /// @dev Register a new public key for the calling Ethereum address
-    /// @param ethAddress The Ethereum address to register
+    /// @dev Register a new Zeto account
     /// @param publicKey The public Babyjubjub key to register
-    function register(
-        address ethAddress,
-        uint256[2] memory publicKey
-    ) public onlyOwner {
-        if (publicKeys[ethAddress][0] != 0 || publicKeys[ethAddress][1] != 0) {
-            revert AlreadyRegistered(ethAddress);
+    function _register(uint256[2] memory publicKey) internal {
+        uint256 nodeHash = _getIdentitiesLeafNodeHash(publicKey);
+        SmtLib.Node memory node = _publicKeysTree.getNode(nodeHash);
+        if (node.nodeType != SmtLib.NodeType.EMPTY) {
+            revert AlreadyRegistered(publicKey);
         }
-        publicKeys[ethAddress] = publicKey;
+        _publicKeysTree.addLeaf(nodeHash, nodeHash);
+    }
+
+    /// @dev returns whether the given public key is registered
+    /// @param publicKey The Babyjubjub public key to check
+    /// @return bool whether the given public key is included in the registry
+    function isRegistered(
```

Comment:
not sure where does this get used.

Reply:
it doesn't need to be used as the circuit does the check with the identitiesRoot https://github.com/hyperledger-labs/zeto/pull/25/files#diff-9383275728425055b13bcbc860ce1ca0d4885b23049fa3ecf52fa71f1f144e65R105

Reply:
yeah it's a convenient method for client apps. not essential for the circuit verifier to work.